drivers/sensor/pcnt_esp32/pcnt_esp32.c

```diff
@@ -0,0 +1,415 @@
+	return 0;
+}
+
+static int pcnt_esp32_attr_get(const struct device *dev, enum sensor_channel chan,
+			       enum sensor_attribute attr, struct sensor_value *val)
+{
+	const struct pcnt_esp32_config *config = dev->config;
+
+	if (chan != SENSOR_CHAN_ALL && chan != SENSOR_CHAN_ROTATION) {
+		return -ENOTSUP;
+	}
+
+	switch (attr) {
+	case SENSOR_ATTR_LOWER_THRESH:
+		val->val1 = config->unit_config[0].l_thr;
+		break;
+	case SENSOR_ATTR_UPPER_THRESH:
+		val->val1 = config->unit_config[0].h_thr;
+		break;
+	case SENSOR_ATTR_OFFSET:
+		val->val1 = config->unit_config[0].offset;
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	val->val2 = 0;
+
+	return 0;
+}
+
+#ifdef CONFIG_PCNT_ESP32_TRIGGER
+static void IRAM_ATTR pcnt_esp32_isr(const struct device *dev)
+{
+	struct pcnt_esp32_data *data = (struct pcnt_esp32_data *const)(dev)->data;
+
+	uint32_t pcnt_intr_status;
+	uint32_t pcnt_unit_status;
+	struct sensor_trigger trigger;
+
+	pcnt_intr_status = pcnt_ll_get_intr_status(data->hal.dev);
+	pcnt_ll_clear_intr_status(data->hal.dev, pcnt_intr_status);
+
+	if (pcnt_intr_status & PCNT_INTR_UNIT_0) {
+		pcnt_unit_status = pcnt_ll_get_unit_status(data->hal.dev, 0);
+	} else if (pcnt_intr_status & PCNT_INTR_UNIT_1) {
+		pcnt_unit_status = pcnt_ll_get_unit_status(data->hal.dev, 1);
+	} else if (pcnt_intr_status & PCNT_INTR_UNIT_2) {
+		pcnt_unit_status = pcnt_ll_get_unit_status(data->hal.dev, 2);
+	} else if (pcnt_intr_status & PCNT_INTR_UNIT_3) {
+		pcnt_unit_status = pcnt_ll_get_unit_status(data->hal.dev, 3);
+#ifdef CONFIG_SOC_ESP32
+	} else if (pcnt_intr_status & PCNT_INTR_UNIT_4) {
+		pcnt_unit_status = pcnt_ll_get_unit_status(data->hal.dev, 4);
+	} else if (pcnt_intr_status & PCNT_INTR_UNIT_5) {
+		pcnt_unit_status = pcnt_ll_get_unit_status(data->hal.dev, 5);
+	} else if (pcnt_intr_status & PCNT_INTR_UNIT_6) {
+		pcnt_unit_status = pcnt_ll_get_unit_status(data->hal.dev, 6);
+	} else if (pcnt_intr_status & PCNT_INTR_UNIT_7) {
+		pcnt_unit_status = pcnt_ll_get_unit_status(data->hal.dev, 7);
+#endif /* CONFIG_SOC_ESP32 */
+	} else {
+		return;
+	}
+
+	if (!(pcnt_unit_status & PCNT_INTR_THRES_0) && !(pcnt_unit_status & PCNT_INTR_THRES_1)) {
+		return;
+	}
+
+	if (!data->trigger_handler) {
+		return;
+	}
+
+	trigger.type = SENSOR_TRIG_THRESHOLD;
+	trigger.chan = SENSOR_CHAN_ROTATION;
+	data->trigger_handler(dev, &trigger);
+}
+
+static int pcnt_esp32_trigger_set(const struct device *dev, const struct sensor_trigger *trig,
+				  sensor_trigger_handler_t handler)
+{
+	int ret;
+	const struct pcnt_esp32_config *config = dev->config;
+	struct pcnt_esp32_data *data = (struct pcnt_esp32_data *const)(dev)->data;
+
+	if (trig->type != SENSOR_TRIG_THRESHOLD) {
+		return -ENOTSUP;
+	}
+
+	if ((trig->chan != SENSOR_CHAN_ALL) && (trig->chan != SENSOR_CHAN_ROTATION)) {
+		return -ENOTSUP;
+	}
+
+	if (!handler) {
+		return -EINVAL;
+	}
+
+	data->trigger_handler = handler;
+
+	ret = esp_intr_alloc(DT_INST_IRQN(0), 0, (intr_handler_t)pcnt_esp32_isr, (void *)dev, NULL);
```

Comment:
That is okay there is only one PCNT controller, but if this driver scale hardcoding here to the instance zero may break the driver.

I suggest to put the irq source into the config structure and pick the irq source using DT macros that instances the driver.

Reply:
It is hardcoded to 0 since the SoCs (esp32 and esp32s2) only have one instance of this peripheral. I will move it to the config structure initialization.